### PR TITLE
r.maxent.predict: a new addon to create prediction layers based on a Maxent model

### PR DESCRIPTION
**Pull request by ecodiv**
This addon complements the r.maxent.train (new, not yet merged) and v.maxent.swd addons

### CODE REVIEW

**Review by wenzeslaus — COMMENTED** [review]
The file `src/raster/r.maxent.predict/tempfile` looks like a mistake from `git add -A`. Awesome images otherwise!

File: src/raster/r.maxent.predict/r.maxent.predict.html [hunk 1]
```diff
@@ -0,0 +1,265 @@
+<h2>DESCRIPTION</h2>
+
+The <em>r.maxent.predict</em> addon can be used to apply a previously 
+calculated Maxent model to a new set of environmental raster data. It 
+requires a <i>.lambdas file</i> file describing a Maxent model, and the
```
Comment (echoix) [inline src/raster/r.maxent.predict/r.maxent.predict.html]:
```suggestion
requires a <i>.lambdas</i> file describing a Maxent model, and the 
```

File: src/raster/r.maxent.predict/r.maxent.predict.html [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+and a number of bioclim raster layers from <a 
+href="https://www.worldclim.org/">WorldClim</a> version 2.1, 
+representing the climate conditions representing the period 1970-2000 
+and the climate conditions predicted for 2061–2080 based on the GCM 
+BCC-CSM2-MR and SSP 585.
+
+
+<p>
+The zip file contains a folder <i>sampledata</i>. This is a location 
+with five sub-folders <i>PERMANENT</i>, <i>sloth</i>, <i>current</i>,
```
Comment (echoix) [inline src/raster/r.maxent.predict/r.maxent.predict.html]:
```suggestion
with five subfolders <i>PERMANENT</i>, <i>sloth</i>, <i>current</i>, 
```

File: src/raster/r.maxent.predict/r.maxent.predict.html [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+ export_rasters=envlayers
+</pre>
+</div>
+
+<p>
+The output is a folder with the so-called SWD files  with the XY 
+coordinates for the species presence location (<i>spec_swd.csv</i>) and 
+the background locations (<i>bgrd_swd.csv</i>. Both also include the 
+values of the input raster layers for the given point locations. In 
+addition, there is the sub-folder <i>envlayers</i> with the
```
Comment (echoix) [inline src/raster/r.maxent.predict/r.maxent.predict.html]:
```suggestion
addition, there is the subfolder <i>envlayers</i> with the 
```

File: src/raster/r.maxent.predict/r.maxent.predict.html [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+coordinates for the species presence location (<i>spec_swd.csv</i>) and 
+the background locations (<i>bgrd_swd.csv</i>. Both also include the 
+values of the input raster layers for the given point locations. In 
+addition, there is the sub-folder <i>envlayers</i> with the 
+environmental raster layers in <i>ascii</i> format.
+
+<h3>2: Train the model</h3>
+
+Use the output of <em>v.maxent.swd</em> as input for 
+<em>rmaxent.train</em>. First create a sub-folder <i>output_model1</i>,
```
Comment (echoix) [inline src/raster/r.maxent.predict/r.maxent.predict.html]:
```suggestion
<em>r.maxent.train</em>. First create a subfolder <i>output_model1</i>, 
```

File: src/raster/r.maxent.predict/r.maxent.predict.html [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+The resulting layer is written to the current mapset as 
+<i>model_1_ssp585</i> (right map in the figure below). The results 
+suggest the area with suitable conditions will increase under future 
+climates compared the that under the current conditions (left map in 
+the figure below). Not sure that is realistic.  
+
+<p>
+<div align="left" style="margin: 10px"> <a 
+href="r_maxent_predict_02.png"> <img src="r_maxent_predict_02.png" 
+alt="Predicted suitabilty map for Bradypus tridactylus for the period
```
Comment (echoix) [inline src/raster/r.maxent.predict/r.maxent.predict.html]:
```suggestion
alt="Predicted suitability map for Bradypus tridactylus for the period 
```

File: src/raster/r.maxent.predict/r.maxent.predict.html [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+In your current mapset, you'll find the raster prediction layer, and 
+the sample and background point layers with the predicted values. 
+
+<p>
+<div align="left" style="margin: 10px"> <a href="r_maxent_train.png"> 
+<img src="r_maxent_train.png" alt="Output layers in grass gis" 
+border="0"> </a><br><i>The example creates the prediction raster layer 
+'model_1_suitability_current',  the sample point layer 
+'model_1_samplepred' and the background point layer 'model_bgrdpred' 
+(for the latter, ony part of the map is shown here).</i> </div>
```
Comment (echoix) [inline src/raster/r.maxent.predict/r.maxent.predict.html]:
```suggestion
(for the latter, only part of the map is shown here).</i> </div>
```

File: src/raster/r.maxent.predict/r.maxent.predict.py [hunk 7]
```diff
@@ -0,0 +1,385 @@
...
+# % description: Names of the raster layers representing the environmental variables used in the Maxent model.
+# % required : no
+# % guisection: input
+# %end
+
+# %option
+# % key: variables
+# % type: string
+# % label: variable names
+# % description: Names of the environmental parameter(s) as used in the model. These need to be given in the same order as the corresponding raster layers. If left out, the names are assumed to be the same as the  name of the raster layers.
```
Comment (echoix) [inline src/raster/r.maxent.predict/r.maxent.predict.py]:
```suggestion
# % description: Names of the environmental parameter(s) as used in the model. These need to be given in the same order as the corresponding raster layers. If left out, the names are assumed to be the same as the name of the raster layers.
```

File: src/raster/r.maxent.predict/r.maxent.predict.py [hunk 8]
```diff
@@ -0,0 +1,385 @@
...
+# %flag
+# % key: i
+# % label: Copy maxent.jar to addon directory
+# % description: Copy the maxent.jar (path provided with the 'maxent' parameter) to the addon scripts directory.
+# %end
+
+# %flag
+# % key: u
+# % label: Overwrites maxent.jar in addon directory
+# % description: Copy the maxent.jar (path provided with the 'maxent' parameter) to the addon scripts directory. If the file already exist in the addon directory, it is overwritten.
```
Comment (echoix) [inline src/raster/r.maxent.predict/r.maxent.predict.py]:
```suggestion
# % description: Copy the maxent.jar (path provided with the 'maxent' parameter) to the addon scripts directory. If the file already exists in the addon directory, it is overwritten.
```

File: src/raster/r.maxent.predict/r.maxent.predict.py [hunk 9]
```diff
@@ -0,0 +1,385 @@
...
+                    " the maxent.jar file, use the -u flag instead."
+                )
+                gs.fatal(_(msg))
+            else:
+                shutil.copyfile(maxent_file, maxent_copy)
+                msg = "Copied the maxent.jar file to the grass gis addon script directory .\n\n"
+                gs.info(_(msg))
+        if bool(flags["u"]):
+            shutil.copyfile(maxent_file, maxent_copy)
+            msg = "Copied the maxent.jar file to the grass gis addon script directory .\n\n"
```
Comment (echoix) [inline src/raster/r.maxent.predict/r.maxent.predict.py]:
```suggestion
            msg = "Copied the maxent.jar file to the GRASS GIS addon script directory .\n\n"
```

File: src/raster/r.maxent.predict/r.maxent.predict.py [hunk 10]
```diff
@@ -0,0 +1,385 @@
...
+import subprocess
+import sys
+import uuid
+import grass.script as gs
+
+
+temp_directory = gs.tempdir()
+
+
+# Funtions
```
Comment (echoix) [inline src/raster/r.maxent.predict/r.maxent.predict.py]:
```suggestion
# Functions
```

File: src/raster/r.maxent.predict/r.maxent.predict.py [hunk 11]
```diff
@@ -0,0 +1,385 @@
...
+            if os.path.isfile(maxent_copy):
+                msg = (
+                    "There is already a maxent.jar file in the scripts \n"
+                    "directory. Remove the -i flag. If you want to update \n"
+                    " the maxent.jar file, use the -u flag instead."
+                )
+                gs.fatal(_(msg))
+            else:
+                shutil.copyfile(maxent_file, maxent_copy)
+                msg = "Copied the maxent.jar file to the grass gis addon script directory .\n\n"
```
Comment (echoix) [inline src/raster/r.maxent.predict/r.maxent.predict.py]:
```suggestion
                msg = "Copied the maxent.jar file to the GRASS GIS addon script directory .\n\n"
```

File: src/raster/r.maxent.predict/r.maxent.predict.py [hunk 12]
```diff
@@ -0,0 +1,385 @@
...
+            gs.info(stderr_line)
+        # Check the return code
+        process.wait()
+        if process.returncode != 0:
+            gs.fatal(_("Maxent terminated with an error"))
+
+    # -----------------------------------------------------------------
+    # Import the resulting layer in GRASS GIS
+    # -----------------------------------------------------------------
+    gs.info(_("Importing the predicted suitability layer in grass gis\n"))
```
Comment (echoix) [inline src/raster/r.maxent.predict/r.maxent.predict.py]:
```suggestion
    gs.info(_("Importing the predicted suitability layer in GRASS GIS\n"))
```

File: src/raster/r.maxent.predict/r.maxent.predict.html [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+href="https://biodiversityinformatics.amnh.org/open_source/maxent/">Maxent 
+website</a>.
+
+<p>
+In your current mapset, you'll find the raster prediction layer, and 
+the sample and background point layers with the predicted values. 
+
+<p>
+<div align="left" style="margin: 10px"> <a href="r_maxent_train.png"> 
+<img src="r_maxent_train.png" alt="Output layers in grass gis"
```
Comment (echoix) [inline src/raster/r.maxent.predict/r.maxent.predict.html]:
```suggestion
<img src="r_maxent_train.png" alt="Output layers in GRASS GIS" 
```

**Review by echoix — COMMENTED** [review]
Typos only review :)

**Comment by ecodiv** [comment]
> Typos only review :)

Thanks! And let me try the batch function this time :-)

**Pull request merged**